what4/src/What4/Utils/AbstractDomains.hs

```diff
@@ -585,14 +600,113 @@ natRangeToRange :: NatValueRange -> ValueRange Integer
+stringAbsJoin (StringAbs lenx) (StringAbs leny) =
+  StringAbs (natJoinRange lenx leny)
+
+stringAbsSingle :: StringLiteral si -> StringAbstractValue
+stringAbsSingle lit = StringAbs (natSingleRange (stringLitLength lit))
+
+stringAbsOverlap :: StringAbstractValue -> StringAbstractValue -> Bool
+stringAbsOverlap (StringAbs lenx) (StringAbs leny) = avOverlap BaseNatRepr lenx leny
+
+stringAbsCheckEq :: StringAbstractValue -> StringAbstractValue -> Maybe Bool
+stringAbsCheckEq (StringAbs lenx) (StringAbs leny)
+  | Just 0 <- asSingleNatRange lenx
+  , Just 0 <- asSingleNatRange leny
+  = Just True
+
+  | not (avOverlap BaseNatRepr lenx leny)
+  = Just False
+
+  | otherwise
+  = Nothing
+
+stringAbsConcat :: StringAbstractValue -> StringAbstractValue -> StringAbstractValue
+stringAbsConcat (StringAbs lenx) (StringAbs leny) =
+  StringAbs (natRangeAdd lenx leny)
+
+stringAbsSubstring :: StringAbstractValue -> NatValueRange -> NatValueRange -> StringAbstractValue
+stringAbsSubstring (StringAbs s) off len = StringAbs (natRange lo hi)
+ where
+ rng = addRange (natRangeToRange s) (negateRange (natRangeToRange off))
+
+ lo  = case rangeLowBound rng of
+         Unbounded -> 0
+         Inclusive x
+           | x < 0 -> 0
+           | otherwise -> min (fromInteger x) (natRangeLow len)
+
+ hi  = case (rangeHiBound rng, natRangeHigh len) of
+         (Unbounded  , y) -> y
+         (Inclusive x, Unbounded)   -> Inclusive (fromInteger (max 0 x))
+         (Inclusive x, Inclusive y) -> Inclusive (min (fromInteger (max 0 x)) y)
+
+
+stringAbsContains :: StringAbstractValue -> StringAbstractValue -> Maybe Bool
+stringAbsContains = couldContain
+
+stringAbsIsPrefixOf :: StringAbstractValue -> StringAbstractValue -> Maybe Bool
+stringAbsIsPrefixOf = flip couldContain
+
+stringAbsIsSuffixOf :: StringAbstractValue -> StringAbstractValue -> Maybe Bool
+stringAbsIsSuffixOf = flip couldContain
+
+couldContain :: StringAbstractValue -> StringAbstractValue -> Maybe Bool
+couldContain (StringAbs lenx) (StringAbs leny)
+  | Inclusive x <- rangeHiBound rng, x < 0 = Just False
+  | otherwise = Nothing
+
+  where
+  lenx' = natRangeToRange lenx
+  leny' = natRangeToRange leny
+
+  rng = addRange lenx' (negateRange leny')
+
+stringAbsIndexOf :: StringAbstractValue -> StringAbstractValue -> NatValueRange -> ValueRange Integer
+stringAbsIndexOf (StringAbs lenx) (StringAbs leny) k
+  | Inclusive x <- rangeHiBound shortRng, x < 0 = SingleRange (-1)
+  | otherwise = MultiRange (Inclusive (-1)) (rangeHiBound rng)
+  where
+  lenx' = natRangeToRange lenx
+  leny' = natRangeToRange leny
+  k'    = natRangeToRange k
+
+  -- possible values that the final offset could have if the substring exists anywhere
+  rng = addRange lenx' (negateRange leny')
+
+  -- compute the values that @off-k@ could have.  If negative,
+  -- the search string could never fit
+  shortRng = addRange lenx' (negateRange (addRange leny' k'))
```

Comment:
As with `couldContain`, maybe the first conditional equation would be easier to understand if written with `natCheckLe`:
```
stringAbsIndexOf (StringAbs lenx) (StringAbs leny) k
  | natCheckLe (natRangeAdd leny k) lenx == False = SingleRange (-1)
  ...
```